### PR TITLE
[OnnxModelLoader] Adding Support of Expand Onnx Operator

### PR DESCRIPTION
**Pull request by psubhasc**
Summary: Many recommendation and Computer vision models when represented in onnx use expand operator, Thus support of this operator is essential

Test Plan: Ninja test was done. Additional Test cases where added for expand operator functionality check.

### CODE REVIEW

**Comment by chadrosier** [comment]
Ping.

File: tests/unittests/OnnxImporterTest.cpp [hunk 1]
```diff
@@ -261,6 +261,67 @@ importArithMultiBroadcastTest(std::string fileName,
...
+                             std::vector<dim_t> outputShape,
+                             std::vector<float> expectedValues) {
+  float delta = 1e-08;
+  ExecutionEngine EE{};
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+  PlaceholderBindings bindings;
+  Placeholder *graphOutputVar;
+  // Load the .onnxtxt model
+  // ONNXModelLoader onnxLD(netFilename, {"x"}, {&x.getType()}, *F);
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: rm this

File: tests/unittests/OnnxImporterTest.cpp [hunk 2]
```diff
@@ -261,6 +261,67 @@ importArithMultiBroadcastTest(std::string fileName,
...
+  Type input_type(ElemKind::FloatTy, inputShape);
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: `inputType`

File: tests/unittests/OnnxImporterTest.cpp [hunk 3]
```diff
@@ -261,6 +261,67 @@ importArithMultiBroadcastTest(std::string fileName,
...
+  ONNXModelLoader onnxLD(netFilename, {"x"}, {&input_type}, *F);
+  graphOutputVar = EXIT_ON_ERR(onnxLD.getSingleOutput());
+  auto PH = mod.getPlaceholderByName("x");
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: `auto *PH`

File: tests/unittests/OnnxImporterTest.cpp [hunk 4]
```diff
@@ -261,6 +261,67 @@ importArithMultiBroadcastTest(std::string fileName,
                                           {bindings.get(graphOutputVar)}));
 }
 
+// Import expand
+static void importExpandTest(std::string &netFilename,
+                             std::vector<float> inputValues,
+                             std::vector<dim_t> inputShape,
+                             std::vector<dim_t> outputShape,
+                             std::vector<float> expectedValues) {
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: const ref here, or `llvm::ArrayRef`

File: tests/unittests/OnnxImporterTest.cpp [hunk 5]
```diff
@@ -261,6 +261,67 @@ importArithMultiBroadcastTest(std::string fileName,
                                           {bindings.get(graphOutputVar)}));
 }
 
+// Import expand
+static void importExpandTest(std::string &netFilename,
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: tiny nit: `const`

File: lib/Importer/ONNXModelLoader.cpp [hunk 6]
```diff
@@ -1670,6 +1679,41 @@ Error ONNXModelLoader::loadTile(const ONNX_NAMESPACE::NodeProto &op,
...
+  auto inputDimSize = (size_t)in.dims().size();
+  auto repeatSize = (size_t)tiles.size();
+  if (repeatSize > inputDimSize) {
+    for (size_t i = 0; i < repeatSize - inputDimSize; i++) {
+      in = G_.createExpandDims(opName + "_" + std::to_string(i), in, i);
+    }
+  }
+
+  Node *N = in;
+  for (size_t i = 0; i < tiles.size(); i++) {
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: nit: We follow LLVM style for `for` loops even when not using an iterator ([see here](https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop)):

`for (size_t i = 0, e = tiles.size(); i < e; i++)`

File: lib/Importer/ONNXModelLoader.cpp [hunk 7]
```diff
@@ -1670,6 +1679,41 @@ Error ONNXModelLoader::loadTile(const ONNX_NAMESPACE::NodeProto &op,
...
+  Constant *repeats;
+  ASSIGN_VALUE_OR_RETURN_ERR(in, getNodeValueByName(op.input(0)));
+  ASSIGN_VALUE_OR_RETURN_ERR(repeats, getConstantByName(op.input(1)));
+
+  std::vector<int64_t> tiles;
+  helperSetter<int64_t, int64_t>(repeats, tiles);
+  auto inputDimSize = (size_t)in.dims().size();
+  auto repeatSize = (size_t)tiles.size();
+  if (repeatSize > inputDimSize) {
+    for (size_t i = 0; i < repeatSize - inputDimSize; i++) {
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: nit: We follow LLVM style for `for` loops even when not using an iterator ([see here](https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop)):

`for (size_t i = 0, e = repeatSize - inputDimSize; i < e; i++)`

File: include/glow/Importer/ONNXModelLoader.h [hunk 8]
```diff
@@ -188,6 +188,10 @@ class ONNXModelLoader
   Error loadTile(const ONNX_NAMESPACE::NodeProto &op,
                  const ArgumentDictionaryTy &dict);
 
+  /// Load Expand ONNX operator
```
Comment (jfix71) [inline include/glow/Importer/ONNXModelLoader.h]: nit: period.

File: tests/unittests/OnnxImporterTest.cpp [hunk 9]
```diff
@@ -261,6 +261,67 @@ importArithMultiBroadcastTest(std::string fileName,
                                           {bindings.get(graphOutputVar)}));
 }
 
+// Import expand
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: Improve comment? And period.

File: lib/Importer/ONNXModelLoader.cpp [hunk 10]
```diff
@@ -692,6 +692,15 @@ static void helperSetter(Constant *constT, std::vector<ssize_t> &vec) {
   }
 }
 
+template <typename T, typename datatype>
+static void helperSetter(const glow::Constant *constT,
```
[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Updated.

File: tests/unittests/OnnxImporterTest.cpp (hunk 1 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 2 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 3 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 4 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 5 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: lib/Importer/ONNXModelLoader.cpp (hunk 6 above)
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Done.

File: lib/Importer/ONNXModelLoader.cpp (hunk 7 above)
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Done.

File: include/glow/Importer/ONNXModelLoader.h (hunk 8 above)
Comment (psubhasc) [reply include/glow/Importer/ONNXModelLoader.h]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 9 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: lib/Importer/ONNXModelLoader.cpp [hunk 11]
```diff
@@ -985,7 +985,7 @@ Error ONNXModelLoader::loadConstant(const ONNX_NAMESPACE::NodeProto &op,
 }
 
 /// Retrieves data from a constant Tensor and stores it in a vector.
-template <typename T>
+template <typename T, typename datatype = ssize_t>
 static void helperSetter(Constant *constT, std::vector<ssize_t> &vec) {
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: Didn't you need to also update vec to use `datatype` here? If this isn't causing any build issues not sure why it's necessary in the first place. Was the idea that there could be a troublesome implicit cast on 32-bit architectures, so this was just to be safe? It'd be better if this was actually exposed somehow in a test, though not sure it's easy.

Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: As of now, I have updated vec to use datatype to be safer side.

**Pull request closed without merging**